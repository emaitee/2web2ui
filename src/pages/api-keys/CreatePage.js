import React from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { Page, Panel } from '@sparkpost/matchbox';

import { createApiKey } from 'src/actions/api-keys';
import Layout from 'src/components/layout/Layout';
import { getLoading } from 'src/selectors/api-keys';
import ApiKeyForm from './components/ApiKeyForm';

const breadcrumbAction = {
  content: 'API Keys',
  Component: Link,
  to: '/account/api-keys'
};

export class CreatePage extends React.Component {
  onSubmit = (values) => {
    const { createApiKey, history } = this.props;

    return createApiKey(values).then((res) => {
      history.push('/account/api-keys');
    });
  };

  render() {
    return (
      <Layout.App loading={this.props.loading}>
        <Page title="Create API Key" breadcrumbAction={breadcrumbAction} />
        <Panel>
          <Panel.Section>
            <ApiKeyForm onSubmit={this.onSubmit} />
          </Panel.Section>
        </Panel>
      </Layout.App>
    );
  }
}

const mapStateToProps = (state, props) => ({
  loading: getLoading(state)
});

export default withRouter(
  connect(mapStateToProps, { createApiKey })(CreatePage)
);
