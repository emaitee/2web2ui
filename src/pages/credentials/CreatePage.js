import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Page, Panel } from '@sparkpost/matchbox';

import { createApiKey } from 'actions/credentials';
import Layout from 'components/layout/Layout';
import { getLoading } from 'selectors/credentials';
import ApiKeyForm from './components/ApiKeyForm';

const breadcrumbAction = {
  content: 'API Keys',
  Component: Link,
  to: '/account/credentials'
};

const CreatePage = ({ createApiKey, loading }) => (
  <Layout.App loading={loading}>
    <Page title="Create API Key" breadcrumbAction={breadcrumbAction} />
    <Panel>
      <Panel.Section>
        <ApiKeyForm onSubmit={createApiKey} />
      </Panel.Section>
    </Panel>
  </Layout.App>
);

const mapStateToProps = (state, props) => ({
  loading: getLoading(state)
});

export default connect(mapStateToProps, { createApiKey })(CreatePage);
