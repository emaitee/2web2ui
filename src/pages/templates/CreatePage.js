import React, { Component } from 'react';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { reduxForm, formValueSelector } from 'redux-form';
// Actions
import { create } from '../../actions/templates';

// Components
import { Layout } from 'src/components';
import Form from './components/Form';
import Editor from './components/Editor';
import { Page, Grid } from '@sparkpost/matchbox';

const FORM_NAME = 'templateCreate';

class CreatePage extends Component {
  state = {
    shouldRedirect: false
  }

  handleCreate(values) {
    const { create } = this.props;
    return create(values)
      .then(() => this.setState({ shouldRedirect: true }));
  }

  renderPageHeader() {
    const { handleSubmit, submitting } = this.props;

    const primaryAction = {
      content: 'Save Template',
      onClick: handleSubmit((values) => this.handleCreate(values)),
      disabled: submitting
    };

    const backAction = {
      content: 'Templates',
      Component: Link,
      to: '/templates'
    };

    return (
      <Page
        primaryAction={primaryAction}
        breadcrumbAction={backAction}
        title='New Template'
      />
    );
  }

  render() {
    const { id } = this.props;

    if (this.state.shouldRedirect) {
      return <Redirect to={`/templates/edit/${id}`} />;
    }

    return (
      <Layout.App>
        { this.renderPageHeader() }
        <Grid>
          <Grid.Column xs={12} lg={4}>
            <Form newTemplate={true} name={FORM_NAME} />
          </Grid.Column>
          <Grid.Column xs={12} lg={8}>
            <Editor name={FORM_NAME} />
          </Grid.Column>
        </Grid>
      </Layout.App>
    );
  }
}

const selector = formValueSelector(FORM_NAME);
const mapStateToProps = (state) => ({
  id: selector(state, 'id')
});

const formOptions = {
  form: FORM_NAME
};

export default connect(mapStateToProps, { create })(reduxForm(formOptions)(CreatePage));
