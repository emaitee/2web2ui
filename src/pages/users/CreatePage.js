import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { Field, reduxForm } from 'redux-form';
import { Page, Panel, Button } from '@sparkpost/matchbox';
import { TextFieldWrapper } from 'src/components';
import RoleSelect from './components/RoleSelect';
import { required, email } from 'src/helpers/validation';
import { inviteUser } from 'src/actions/users';
import { showAlert } from 'src/actions/globalAlert';

const formName = 'userForm';
const breadcrumbAction = {
  content: 'Users',
  Component: Link,
  to: '/account/users'
};

export class CreatePage extends Component {

  handleSubmit = (values) => {
    const {
      inviteUser,
      showAlert,
      history
    } = this.props;
    const { email, access } = values;

    return inviteUser(email, access)
      .then(() => {
        showAlert({
          type: 'success',
          message: `Invitation sent to ${email}`
        });
        history.push('/account/users');
      })
      .catch((err) => {
        const { response: { status, data }} = err;
        let message = 'Unable to invite user.';

        if (status === 400) {
        // Email address suppressed
          message = 'The email you tried to invite is currently suppressed by SparkPost. Please use another email address or contact support.';
        } else if (status === 409) {
        // User already exists
          message = data.errors[0].message;
        }

        showAlert({
          type: 'error',
          message
        });
      });
  };

  render() {
    const {
      submitting,
      pristine,
      handleSubmit
    } = this.props;

    return (
      <Page title="Add User" breadcrumbAction={breadcrumbAction}>
        <Panel>
          <Panel.Section>
            <form onSubmit={handleSubmit(this.handleSubmit)}>
              <p>An invitation will be sent to the email address you supply</p>
              <Field
                name="email"
                validate={[required, email]}
                label="Email address"
                component={TextFieldWrapper}
              />
              <Field name="access" label="Role" component={RoleSelect} />
              <Button submit primary disabled={submitting || pristine}>
                {submitting ? 'Loading' : 'Add User' }
              </Button>
            </form>
          </Panel.Section>
        </Panel>
      </Page>
    );
  }

}

const mapStateToProps = () => ({
  initialValues: {
    access: 'admin' // Sadly redux-form does not reflect a select's initial value
  }
});

const mapDispatchToProps = { inviteUser, showAlert };

const ReduxCreatePage = reduxForm({ form: formName })(CreatePage);

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(ReduxCreatePage)
);
