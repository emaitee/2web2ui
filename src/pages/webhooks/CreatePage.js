import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import _ from 'lodash';

// Actions
import { createWebhook, getEventDocs } from '../../actions/webhooks';

// Components
import { Layout } from 'src/components';
import { Page, Panel } from '@sparkpost/matchbox';
import WebhookForm from './components/WebhookForm';

class WebhooksCreate extends Component {
  state = {}

  componentDidMount() {
    if (!this.props.eventDocs) {
      this.props.getEventDocs();
    }
  }

  /*
    Makes a webhook object from form values and calls the createWebhook action
    with it. Invoked in the form's onSubmit func
  */
  createWebhook(values, eventsTree) {
    const { name, target, eventsRadio, auth } = values;

    const webhook = {
      name: name,
      target: target
    };

    let events;

    if (eventsRadio === 'select') {
      const checkedEvents = _.concat(values.message_event, values.track_event, values.gen_event, values.unsubscribe_event, values.relay_event);
      events = _.filter(checkedEvents, (event) => (event)); // filter for truthy
    } else {
      // all events
      events = _.flatten(_.map(eventsTree, ({ events }) => _.map(events, ({ key }) => (key))));
    }

    webhook.events = events;

    // builds the webhooks auth details from the form values
    switch (auth) {
      case 'basic':
        webhook.auth_type = 'basic';
        webhook.auth_credentials = {
          username: values.basicUser,
          password: values.basicPass
        };
        break;
      case 'oauth2':
        webhook.auth_type = 'oauth2';
        webhook.auth_request_details = {
          url: values.tokenURL,
          body: {
            client_id: values.clientId,
            client_secret: values.clientSecret
          }
        };
        break;
      default: // none
        break;
    }

    return this.props.createWebhook(webhook).then(() => this.setState({ redirectTo: '/webhooks' }));
  }

  /*
    Builds a tree of event data, based on the eventDocs, for the form to create
    the checkbox groups with.
  */
  buildEventsData(eventGroups) {
    return _.map(eventGroups, ({ display_name, description, events }, key) => ({
      key: key,
      label: display_name,
      description: description,
      events: _.map(events, ({ display_name, description }, eventKey) => ({
        key: eventKey,
        label: display_name,
        description: description
      }))
    }));
  }

  render() {
    const { redirectTo } = this.state;

    if (redirectTo) {
      return <Redirect to={redirectTo} />;
    }

    const { eventsLoading } = this.props;  // Form doesn't load until we have events
    const { eventDocs } = this.props;
    const eventsTree = this.buildEventsData(eventDocs);

    return (
      <Layout.App loading={eventsLoading}>
        <Page
          title='Create Webhook'
          breadcrumbAction={{ content: 'Webhooks', Component: Link, to: '/webhooks' }}
        />
        <Panel>
          <Panel.Section>
            <WebhookForm eventsTree={eventsTree} newWebhook={true} onSubmit={(values) => this.createWebhook(values, eventsTree)}/>
          </Panel.Section>
        </Panel>
      </Layout.App>
    );
  }
}

const mapStateToProps = ({ webhooks }) => ({
  eventsLoading: webhooks.docsLoading,
  eventDocs: webhooks.docs
});

export default connect(mapStateToProps, { createWebhook, getEventDocs })(WebhooksCreate);
