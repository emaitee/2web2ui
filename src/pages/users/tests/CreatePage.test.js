import React from 'react';
import { shallow } from 'enzyme';
import { CreatePage } from '../CreatePage';

describe('Page: User Create Page', () => {

  let props;
  let wrapper;
  let instance;
  let values;

  beforeEach(() => {
    props = {
      submitting: false,
      pristine: false,
      handleSubmit: jest.fn((fn) => fn),
      inviteUser: jest.fn(() => Promise.resolve()),
      showAlert: jest.fn(),
      history: {
        push: jest.fn()
      }
    };
    wrapper = shallow(<CreatePage {...props} />);
    instance = wrapper.instance();
    values = { email: 'email', access: 'access' };
  });

  it('should render correctly by default', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render when submitting', () => {
    wrapper.setProps({ submitting: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render when pristine', () => {
    wrapper.setProps({ pristine: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should invite a user on submit', async() => {
    await instance.handleSubmit(values);

    expect(props.inviteUser).toHaveBeenCalledWith(values.email, values.access);
    expect(props.showAlert).toHaveBeenCalledWith({
      type: 'success',
      message: `Invitation sent to ${values.email}`
    });
    expect(props.history.push).toHaveBeenCalledWith('/account/users');
  });

  it('should show an alert on a 400 error', async() => {
    const err = { response: { status: 400 }};
    props.inviteUser = jest.fn(() => Promise.reject(err));
    props.inviteUser.mockName('inviteUserMock');

    await instance.handleSubmit(values);

    expect(props.inviteUser).toHaveBeenCalledWith(values.email, values.access);
    expect(props.showAlert).toHaveBeenCalledWith({
      type: 'error',
      message: 'The email you tried to invite is currently suppressed by SparkPost. Please use another email address or contact support.'
    });
    expect(props.history.push).not.toHaveBeenCalled();
  });

  it('should show an alert on a 409 error', () => {

  });

  it('should show an alert on a 500 error', () => {

  });

});
