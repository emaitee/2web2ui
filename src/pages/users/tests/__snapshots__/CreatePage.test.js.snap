// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page: User Create Page should render correctly by default 1`] = `
<Page
  breadcrumbAction={
    Object {
      "Component": [Function],
      "content": "Users",
      "to": "/account/users",
    }
  }
  empty={Object {}}
  title="Add User"
>
  <Panel>
    <Panel.Section>
      <form
        onSubmit={[Function]}
      >
        <p>
          An invitation will be sent to the email address you supply
        </p>
        <Field
          component={[Function]}
          label="Email address"
          name="email"
          validate={
            Array [
              [Function],
              [Function],
            ]
          }
        />
        <Field
          component={[Function]}
          label="Role"
          name="access"
        />
        <Button
          disabled={false}
          primary={true}
          size="default"
          submit={true}
        >
          Add User
        </Button>
      </form>
    </Panel.Section>
  </Panel>
</Page>
`;

exports[`Page: User Create Page should render when pristine 1`] = `
<Page
  breadcrumbAction={
    Object {
      "Component": [Function],
      "content": "Users",
      "to": "/account/users",
    }
  }
  empty={Object {}}
  title="Add User"
>
  <Panel>
    <Panel.Section>
      <form
        onSubmit={[Function]}
      >
        <p>
          An invitation will be sent to the email address you supply
        </p>
        <Field
          component={[Function]}
          label="Email address"
          name="email"
          validate={
            Array [
              [Function],
              [Function],
            ]
          }
        />
        <Field
          component={[Function]}
          label="Role"
          name="access"
        />
        <Button
          disabled={true}
          primary={true}
          size="default"
          submit={true}
        >
          Add User
        </Button>
      </form>
    </Panel.Section>
  </Panel>
</Page>
`;

exports[`Page: User Create Page should render when submitting 1`] = `
<Page
  breadcrumbAction={
    Object {
      "Component": [Function],
      "content": "Users",
      "to": "/account/users",
    }
  }
  empty={Object {}}
  title="Add User"
>
  <Panel>
    <Panel.Section>
      <form
        onSubmit={[Function]}
      >
        <p>
          An invitation will be sent to the email address you supply
        </p>
        <Field
          component={[Function]}
          label="Email address"
          name="email"
          validate={
            Array [
              [Function],
              [Function],
            ]
          }
        />
        <Field
          component={[Function]}
          label="Role"
          name="access"
        />
        <Button
          disabled={true}
          primary={true}
          size="default"
          submit={true}
        >
          Loading
        </Button>
      </form>
    </Panel.Section>
  </Panel>
</Page>
`;
